server/asset/common.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2019, The Decred developers
+// See LICENSE for details.
+
+package asset
+
+import "github.com/decred/slog"
+
+type Error string
+
+func (err Error) Error() string { return string(err) }
+
+const (
+	UnsupportedScriptError = Error("unsupported script type")
+)
+
+// Network flags passed to asset backends to signify which network to use.
+type Network uint8
+
+const (
+	Mainnet Network = iota
+	Testnet
+	Regtest
+)
+
+// The DEX recognizes only three networks. Simnet is a alias of Regtest.
+const Simnet = Regtest
+
+// Every backend constructor will accept a Logger. All logging should take place
+// through the provided logger.
+type Logger = slog.Logger
+
+// The DEXAsset interface is an interface for a blockchain backend. The DEX
+// primarily needs to track UTXOs and transactions to validate orders, and
+// to monitor community conduct.
+type DEXAsset interface {
+	// UTXO should return a UTXO only for outputs that would be spendable on the
+	// blockchain immediately. Pay-to-script-hash UTXOs require the redeem script
+	// in order to calculate sigScript length and verify pubkeys.
+	UTXO(txid string, vout uint32, redeemScript []byte) (UTXO, error)
+	// BlockChannel creates and returns a new channel on which to receive updates
+	// when new blocks are connected.
+	BlockChannel(size int) chan uint32
+	// Transaction returns a DEXTx, which has methods for checking UTXO spending
+	// and swap contract info.
+	Transaction(txid string) (DEXTx, error)
+	// InitTxSize is the size of a serialized atomic swap initialization
+	// transaction with 1 input spending a P2PKH utxo, 1 swap contract output and
+	// 1 change output.
+	InitTxSize() uint32
+}
+
+// UTXO provides data about an unspent transaction output.
+type UTXO interface {
+	// Confirmations returns the number of confirmations for a UTXO's transaction.
+	// Because a UTXO can become invalid after once being considered valid, this
+	// condition should be checked for during confirmation counting and an error
+	// returned if this UTXO is no longer ready to spend. An unmined transaction
+	// should have zero confirmations. A transaction in the current best block
+	// should have one confirmation. A negative number can be returned if error
+	// is not nil.
+	Confirmations() (int64, error)
+	// PaysToPubkeys checks that the provided pubkeys can spend the UTXO and
+	// the signatures are valid.
+	PaysToPubkeys(pubkeys, sigs [][]byte, msg []byte) (bool, error)
```

Comment:
Combined the signature checking and pubkey checking into the single method. Maybe it should be named something else to convey the dual functionality better. Maybe `Auth`?

Reply:
One thing that's a little imprecise in the comment is "the provided pubkeys can spend the UTXO", since really it's closer to "the owner of the pubkeys can spend the UTXO" since the implication is that the corresponding private key(s) can spend the UTXO... the same private key(s) being used to make the signatures.

Anyway, I'm just talking through it to settle on a more meaningful function name.  If we neglect the UTXO part, this would just be `VerifySig` (or whatever).  Since the UTXO part just makes sure that the pubkeys are relevant to the UTXO at hand, the function docs could just say that false is also returned if the UTXO does not pay to the pubkeys even if the signature is valid for the message.  Otherwise the function name could be `PayToPubkeysSigningMessage` (sort of kidding there).  Honestly not sure which way to go here.